state/hook.go

```diff
@@ -0,0 +1,107 @@
+package state
+
+import (
+	"fmt"
+	"github.com/roboll/helmfile/environment"
+	"github.com/roboll/helmfile/helmexec"
+	"github.com/roboll/helmfile/tmpl"
+	"os"
+)
+
+const (
+	ReleaseEventHookPrepare = "prepare"
+	ReleaseEventHookCleanup = "cleanup"
+)
+
+type ReleaseEventHookSpec struct {
+	Name    string   `yaml:"name"`
+	Events  []string `yaml:"events"`
+	Command []string `yaml:"command"`
+	Args    []string `yaml:"args"`
+}
+
+type ReleaseEventHookTemplateData struct {
+	Release         *ReleaseSpec
+	HelmfileCommand string
+	// Environment is accessible as `.Environment` from any template executed by the renderer
+	Environment environment.Environment
+	// Namespace is accessible as `.Namespace` from any non-values template executed by the renderer
+	Namespace string
+	Event     Event
+}
+
+type Event struct {
+	Name string
+}
+
+func (state *HelmState) triggerPrepareEvent(r *ReleaseSpec, helmfileCommand string) (bool, error) {
+	return state.triggerEvent(r, helmfileCommand, ReleaseEventHookPrepare)
+}
+
+func (state *HelmState) triggerCleanupEvent(r *ReleaseSpec, helmfileCommand string) (bool, error) {
+	return state.triggerEvent(r, helmfileCommand, ReleaseEventHookCleanup)
+}
+
+func (state *HelmState) triggerEvent(r *ReleaseSpec, helmfileCommand string, event string) (bool, error) {
+	if state.runner == nil {
+		state.runner = helmexec.ShellRunner{
+			Dir: state.basePath,
+		}
+	}
+
+	for _, hook := range r.Hooks {
+		contained := false
+		for _, e := range hook.Events {
+			contained = contained || e == event
+		}
+		if !contained {
+			continue
+		}
+
+		var err error
+
+		name := hook.Name
+		if name == "" {
+			name = hook.Command[0]
+		}
+
+		fmt.Fprintf(os.Stderr, "%s: basePath=%s\n", state.FilePath, state.basePath)
+		render := tmpl.NewTextRenderer(state.readFile, state.basePath, ReleaseEventHookTemplateData{
+			Release:         r,
+			HelmfileCommand: helmfileCommand,
+			Environment:     state.Env,
+			Namespace:       state.Namespace,
+			Event: Event{
+				Name: event,
+			},
+		})
+
+		state.logger.Debugf("hook[%s]: triggered by event \"%s\"\n", name, event)
+
+		command := make([]string, len(hook.Command))
+		args := make([]string, len(hook.Args))
+		for i, raw := range hook.Command {
+			command[i], err = render.RenderTemplateText(raw)
+			if err != nil {
+				return false, fmt.Errorf("hook[%s]: %v", name, err)
+			}
+		}
+		for i, raw := range hook.Args {
+			args[i], err = render.RenderTemplateText(raw)
+			if err != nil {
+				return false, fmt.Errorf("hook[%s]: %v", name, err)
+			}
+		}
+
+		allargs := []string{}
+		allargs = append(allargs, command[1:]...)
+		allargs = append(allargs, args...)
+		bytes, err := state.runner.Execute(command[0], allargs)
```

Comment:
command is a slice, but only the first one is executed. How does that work if the user specifies multiple commands ?

```
command: ["echo", "helm"]
args: ["hello", "world", "list"]
```

What do you think about these UIs?:

```
command: ["command", "arg1", "arg2"]
```
or
```
commands:
  - cmd: "helm"
    args: "list"
```

Reply:
Trailing items in the command array is considerd args. I made it so to make it consistent with k8s pod spec, hoping it is easy to remember along with it.

Reply:
OK, but the podSpec is [meant to override CMD and ENTRYPOINTS of containers](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#notes). Not sure the model applies. Your call :-)

Reply:
Good point! Maybe blindly trying to make it consistent with pod spec doesn't make sense. I'll just make it a singular value. Thanks a lot for your comments 👍